pkg/grpc/labelstore/client/client.go

```diff
@@ -51,15 +59,31 @@ func (c Client) WatchMatches(selector klabels.Selector, labelType labels.Type, q
 		for {
 			labeled, err := watchClient.Recv()
 			if grpc.Code(err) == codes.Canceled {
+				c.logger.Infoln("label store client: terminating WatchMatches()")
 				// This just means quitCh fired and the RPC was canceled as expected
 				return
 			}
 
 			if err != nil {
-				// doesn't hurt to call cancelFunc() more than once potentially
-				cancelFunc()
-				c.logger.WithError(err).Errorln("Unexpected error reading from WatchMatches stream")
-				return
+				innerCancel()
+				c.logger.WithError(err).Errorln("unexpected error reading from WatchMatches stream, starting another RPC")
+
+				watchClient = nil
+
+				for watchClient == nil {
+					innerCtx, innerCancel = context.WithCancel(outerCtx)
+
+					time.Sleep(2 * time.Second)
```

Comment:
We certainly do want an exponential backoff, however calling `WatchMatches` does not actually initiate a TCP connection to a server. `grpc.Dial()` (the output of which is the `grpc.ClientConn` we get in the `Client` constructor) handles reestablishing TCP connections and it uses exponential backoff. All i'm trying to prevent here is a CPU-sucking busy-loop.